README.md
# Nap

Sleeping for a short time can be relaxing.

[License: MIT](https://git.sr.ht/~sthagen/nap/tree/default/item/LICENSE)

## Documentation

You can create a man page per:

```console
❯ make man
... man share/man/man1/nap.1
```

### Synopsis

```console
nap [base [variation]]
```

### Help

```console
❯ ./nap help
Nap implements real sleep (offering sub-second busy waits). The first real
value argument (base) in seconds determines the base duration. When given a real
number as second argument (variation) the duration will vary uniform randomly
plus minus the 2nd argument.

Default mode just sleeps for a second. This behavior is identical to
when given a single argument of 1.

Usage:

    nap [base [variation]]

The flags are:

    -h
        display help message

    -v
        display the version string

Examples:

❯ nap 0.75 0.1 # sleeps for a randomly selected duration in [0.65, 0.85] seconds

Caveat emptor: Maybe this command is not useful for your use cases.
```

### Version

```console
> ./nap version
v2023.9.20
```


### Building

```console
❯ make
... building for linux/amd64, local, and windows/amd64
... smoke test of local app
v2023.9.20
Nap implements real sleep (offering sub-second busy waits). The first real
value argument (base) in seconds determines the base duration. When given a real
number as second argument (variation) the duration will vary uniform randomly
plus minus the 2nd argument.

Default mode just sleeps for a second. This behavior is identical to
when given a single argument of 1.

Usage:

    nap [base [variation]]

The flags are:

    -h
        display help message

    -v
        display the version string

Examples:

❯ nap 0.75 0.1 # sleeps for a randomly selected duration in [0.65, 0.85] seconds

Caveat emptor: Maybe this command is not useful for your use cases.
... man share/man/man1/nap.1
PASS
coverage: 98.0% of statements
ok    nap 0.304s
nap/main.go:77:   ParseFloat    100.0%
nap/main.go:87:   ParseBase   100.0%
nap/main.go:98:   ParseVariation    100.0%
nap/main.go:108:  RandomDuration    100.0%
nap/main.go:115:  HelpRequested   100.0%
nap/main.go:126:  VersionRequested  100.0%
nap/main.go:137:  HandleAnyErrors   100.0%
nap/main.go:145:  Execute     100.0%
nap/main.go:176:  Seed      100.0%
nap/main.go:181:  main      0.0%
total:      (statements)    98.0%
... open coverage.html
```

### Explorative Testing

Benchmarking the default scenario with [hyperfine](https://crates.io/crates/hyperfine) against some sleep binary a random machine provides:

```console
❯ hyperfine --warmup 1 'sleep 1' './nap 1'
Benchmark 1: sleep 1
  Time (mean ± σ):      1.014 s ±  0.004 s    [User: 0.002 s, System: 0.004 s]
  Range (min … max):    1.009 s …  1.020 s    10 runs

Benchmark 2: ./nap 1
  Time (mean ± σ):      1.014 s ±  0.002 s    [User: 0.003 s, System: 0.003 s]
  Range (min … max):    1.010 s …  1.017 s    10 runs

Summary
  sleep 1 ran
    1.00 ± 0.00 times faster than ./nap-local 1
```

Testing another scenario requesting naps between 0.5 and 1.0 seconds:

```console
❯ hyperfine './nap .75 .25'
Benchmark 1: ./nap .75 .25
  Time (mean ± σ):     748.5 ms ± 153.2 ms    [User: 3.0 ms, System: 3.6 ms]
  Range (min … max):   528.4 ms … 979.4 ms    10 runs
```

## Bug Tracker

Any feature requests or bug reports shall go to the [todos of nap](https://todo.sr.ht/~sthagen/nap).

## Primary Source repository

The main source of `nap` is on a mountain in central Switzerland.
We use distributed version control (git).
There is no central hub.
Every clone can become a new source for the benefit of all.
The preferred public clones of `nap` are:

* [on codeberg](https://codeberg.org/sthagen/nap) - a democratic community-driven, non-profit software development platform operated by Codeberg e.V.
* [at sourcehut](https://git.sr.ht/~sthagen/nap) - a collection of tools useful for software development.

## Contributions

Please do not submit "pull requests" (I found no way to disable that "feature" on GitHub).
If you like to share small changes under the repositories license please kindly do so by sending a patchset.
You can either send such a patchset per email using [git send-email](https://git-send-email.io) or
if you are a sourcehut user by selecting "Prepare a patchset" on the summary page of your fork at [sourcehut](https://git.sr.ht/).

## Status

Prototyping.

**Note**: The default branch is `default`.


docs/changes.md
# Changes

2023.9.20
:    * Added a changes file (<https://todo.sr.ht/~sthagen/nap/3>)
* Added a man page (<https://todo.sr.ht/~sthagen/nap/4>)

2023.9.19
:    * Added a version info to the code (<https://todo.sr.ht/~sthagen/nap/2>)
* Simplified help request detection (<https://todo.sr.ht/~sthagen/nap/2>)

2023.9.18
:    * Initial version


main_test.go
package main

import (
	"bytes"
	"errors"
	"math"
	"os"
	"os/exec"
	"testing"
)

// TestParseBaseValid calls nap.ParseBase with a valid value.
func TestParseBaseValid(t *testing.T) {
	text := "3.1415"
	want := 3.1415
	base, err := ParseBase(text)
	if base != want || err != nil {
		t.Fatalf(`ParseBase("3.1415") = %v, %v, want match for %v, nil`, base, err, want)
	}
}

// TestParseBaseWrongType calls nap.ParseBase with a non-convertible value.
func TestParseBaseWrongType(t *testing.T) {
	text := "foo"
	want := math.NaN()
	msg := errors.New(USAGE + " - failed to parse the base duration")
	base, err := ParseBase(text)
	if !math.IsNaN(base) || err == nil {
		t.Fatalf(`ParseBase("foo") = %v, %v, want match for %v, %v`, base, err, want, msg)
	}
}

// TestParseVariationValid calls nap.ParseVariation with valid values.
func TestParseVariationValid(t *testing.T) {
	text := "1"
	want := 1.0
	base := 3.1415
	variation, err := ParseVariation(text, base)
	if variation != want || err != nil {
		t.Fatalf(`ParseVariation("1", 3.1415) = %v, %v, want match for %v, nil`, base, err, want)
	}
}

// TestParseVariationWrongType calls nap.ParseVariation with a non-convertible value for variation.
func TestParseVariationWrongType(t *testing.T) {
	text := "foo"
	want := math.NaN()
	base := 3.1415
	msg := errors.New(USAGE + " - failed to parse the variation of duration")
	variation, err := ParseVariation(text, base)
	if !math.IsNaN(variation) || err == nil {
		t.Fatalf(`ParseVariation("foo", 3.1415) = %v, %v, want match for %v, %v`, variation, err, want, msg)
	}
}

// TestRandomDurationZeroVariation calls nap.RandomDuration with valid values and corner case of variation zero (0).
func TestRandomDurationZeroVariation(t *testing.T) {
	base := 3.1415
	variation := 0.0
	want := base
	duration := RandomDuration(base, variation)
	if duration != want {
		t.Fatalf(`RandomDuration(3.1415, 0.0) = %v, want match for %v`, duration, want)
	}
}

// TestHelpRequested calls nap.HelpRequested with a match in the list
func TestHelpRequested(t *testing.T) {
	args := []string{"1", "help"}
	helpDetected := HelpRequested(args)
	if !helpDetected {
		t.Fatalf(`HelpRequested(["1", "help"]) = %v, %v, want match for %v`, helpDetected, args, true)
	}
}

// TestHelpRequested calls nap.HelpRequested with no match in the list
func TestHelpRequestedNone(t *testing.T) {
	args := []string{"1", "2", "a", "b", "c", "d"}
	helpDetected := HelpRequested(args)
	if helpDetected {
		t.Fatalf(`HelpRequested(["1", "2", "a", "b", "c", "d"]) = %v, %v, want match for %v`, helpDetected, args, false)
	}
}

// TestVersionRequested calls nap.VersionRequested with a match in the list
func TestVersionRequested(t *testing.T) {
	args := []string{"1", "v"}
	versionDetected := VersionRequested(args)
	if !versionDetected {
		t.Fatalf(`VersionRequested(["1", "v"]) = %v, %v, want match for %v`, versionDetected, args, true)
	}
}

// TestVersionRequestedNone calls nap.VersionRequested with no match in the list
func TestVersionRequestedNone(t *testing.T) {
	args := []string{"1", "hel", "/h", "--Help", "-H", "?"}
	versionDetected := VersionRequested(args)
	if versionDetected {
		t.Fatalf(`VersionRequested(["1", "hel", "/h", "--Help", "-H", "?"]) = %v, %v, want match for %v`, versionDetected, args, false)
	}
}

// TestHandleAnyErrorsNone calls nap.HandleAnyErrors with no error so no output and simple return.
func TestHandleAnyErrorsNone(t *testing.T) {
	tt := []struct {
		description string
		error       error
		expect      string
		code        int
	}{
		{"empty", nil, "", 0},
	}

	for _, tc := range tt {
		t.Run(tc.description, func(t *testing.T) {
			var output bytes.Buffer
			HandleAnyErrors(&output, tc.error, tc.code)
			if tc.expect != output.String() {
				t.Errorf("got %s but expected %s", output.String(), tc.expect)
			}
		})
	}
}

// TestHandleAnyErrorsSome calls nap.HandleAnyErrors with an error so expect output and process exit with code.
func TestHandleAnyErrorsSome(t *testing.T) {
	error := errors.New("An error.")
	want := "An error.\n"
	code := 2

	// Run the crashing code when FLAG is set
	if os.Getenv("FLAG") == "1" {
		var output bytes.Buffer
		HandleAnyErrors(&output, error, code)
		if want != output.String() {
			t.Errorf("got %s but expected %s", output.String(), want)
		}
		return
	}
	// Run the test in a subprocess
	cmd := exec.Command(os.Args[0], "-test.run=TestHandleAnyErrorsSome")
	cmd.Env = append(os.Environ(), "FLAG=1")
	err := cmd.Run()

	// Cast the error as *exec.ExitError and compare the result
	e, ok := err.(*exec.ExitError)
	expectedErrorString := "exit status 2"
	if expectedErrorString != e.Error() {
		t.Errorf("got %s but expected %s", e.Error(), expectedErrorString)
	}
	if !ok {
		t.Errorf("got %v but expected %v", ok, true)
	}
}

// TestExecute calls nap.Execute with valid numeric arguments.
func TestExecute(t *testing.T) {
	args := []string{"0.2", "0.1"}
	var output bytes.Buffer
	code := Execute(&output, 42, args)
	if code != 0 {
		t.Errorf(`Execute(["0.2", "0.1"]) = %v, want match for %v"`, code, 0)
	}
	if "" != output.String() {
		t.Errorf("got %s but expected ''", output.String())
	}
}

// TestExecuteHelp calls nap.Execute with a valid help argument.
func TestExecuteHelp(t *testing.T) {
	args := []string{"--help"}
	var output bytes.Buffer
	code := Execute(&output, 0, args)
	if code != 0 {
		t.Errorf(`Execute(["--help"]) = %v, want match for %v"`, code, 0)
	}
	if HELP+"\n" != output.String() {
		t.Errorf("got %s but expected 'the help text'", output.String())
	}
}

// TestExecuteVersion calls nap.Execute with a valid version argument.
func TestExecuteVersion(t *testing.T) {
	args := []string{"version"}
	var output bytes.Buffer
	code := Execute(&output, 0, args)
	if code != 0 {
		t.Errorf(`Execute(["version"]) = %v, want match for %v"`, code, 0)
	}
	if VERSION+"\n" != output.String() {
		t.Errorf("got %s but expected 'the version text'", output.String())
	}
}

// TestSeed calls nap.Seed with a constant argument.
func TestSeed(t *testing.T) {
	Seed(42)
}


main.go
/*
Nap implements real sleep (offering sub-second busy waits).
The first real value argument (base) in seconds determines the base duration.
When given a real number as second argument (variation)
the duration will vary uniform randomly plus minus the 2nd argument.

Default mode just sleeps for a second.
This behavior is identical to when given a single argument of 1.

Usage:

	nap [base [variation]]

The flags are:

	-h
	    display help message

	-v
	    display the version string

Examples:

❯ nap 0.75 0.1  # sleeps for a randomly selected duration in [0.65, 0.85] seconds

Caveat emptor: Maybe this command is not useful for your use cases.
*/
package main

import (
	"errors"
	"fmt"
	"io"
	"math"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

const (
	DEFAULT_SECS = 1.0 // Base value in seconds When no arguments are given
	HELP         = `Nap implements real sleep (offering sub-second busy waits). The first real
value argument (base) in seconds determines the base duration. When given a real
number as second argument (variation) the duration will vary uniform randomly
plus minus the 2nd argument.

Default mode just sleeps for a second. This behavior is identical to
when given a single argument of 1.

Usage:

    nap [base [variation]]

The flags are:

    -h
        display help message

    -v
        display the version string

Examples:

❯ nap 0.75 0.1 # sleeps for a randomly selected duration in [0.65, 0.85] seconds

Caveat emptor: Maybe this command is not useful for your use cases.` // update after go doc nap shows different help
	MILLIS  = 1000 // Rescaling to adapt to the time.Duration domain
	USAGE   = "usage: nap [base [variation]]"
	VERSION = "v2023.9.20"
)

// ParseFloat parses the given text as positive float and returns code, message, and parsed value as tuple.
// The scope value enables the caller to amend the message with context information.
// If the conversion fails, the value is set to Not-a-Number (NaN).
func ParseFloat(text string, scope string) (int, string, float64) {
	value, err := strconv.ParseFloat(text, 64)
	if err != nil {
		return 2, fmt.Sprintf(USAGE + scope), math.NaN()
	}
	return 0, "", math.Max(0.0, value)
}

// ParseBase attempts to parse the base duration.
// If succeeding, returns the value and nil, else returns NaN and the error.
func ParseBase(text string) (float64, error) {
	code, message, base := ParseFloat(text, " - failed to parse the base duration")
	if code != 0 {
		return math.NaN(), errors.New(message)
	}
	return base, nil
}

// ParseVariation attempts to parse the variation in the context of base.
// If succeeding, ensures the variation is less than or equal to base and returns the value and nil,
// else returns NaN and the error.
func ParseVariation(text string, base float64) (float64, error) {
	code, message, variation := ParseFloat(text, " - failed to parse the variation of duration")
	if code != 0 {
		return math.NaN(), errors.New(message)
	}
	variation = math.Min(base, variation)
	return variation, nil
}

// RandomDuration derives a random duration in seconds given a base and the variation.
func RandomDuration(base float64, variation float64) float64 {
	span := variation + variation
	center := base - variation
	return rand.Float64()*span + center
}

// HelpRequested searches for typical help requests in the command arguments and returns the result.
func HelpRequested(args []string) bool {
	for _, cmd := range args {
		cmdLower := strings.ToLower(cmd)
		if strings.Contains(cmdLower, "h") || strings.Contains(cmdLower, "?") {
			return true
		}
	}
	return false
}

// VersionRequested searches for typical version requests in the command arguments and returns the result.
func VersionRequested(args []string) bool {
	for _, cmd := range args {
		cmdLower := strings.ToLower(cmd)
		if strings.Contains(cmdLower, "v") {
			return true
		}
	}
	return false
}

// HandleAnyErrors prints out the error and exits with code if err is not nil else does nothing.
func HandleAnyErrors(w io.Writer, err error, code int) {
	if err != nil {
		fmt.Fprintln(w, err)
		os.Exit(code)
	}
}

// Execute implements a testable main function.
func Execute(w io.Writer, seed int64, args []string) int {
	if len(args) > 0 && HelpRequested(args) {
		fmt.Fprintln(w, HELP)
		return 0
	}
	if len(args) > 0 && VersionRequested(args) {
		fmt.Fprintln(w, VERSION)
		return 0
	}

	if seed != 0 {
		Seed(seed)
	}
	duration := DEFAULT_SECS
	if len(args) > 0 {
		base, err := ParseBase(args[0])
		HandleAnyErrors(w, err, 2)
		duration = base // duration if no variation requested
		if len(args) > 1 {
			variation, err := ParseVariation(args[1], base)
			HandleAnyErrors(w, err, 2)
			duration = RandomDuration(base, variation) // duration adapted to variation
		}

	}

	time.Sleep(time.Duration(MILLIS*duration) * time.Millisecond)
	return 0
}

// Seed the random number generator.
func Seed(seed int64) {
	rand.Seed(seed)
}

// main perceives the request and acts on it.
func main() {
	os.Exit(Execute(os.Stdout, time.Now().UnixNano(), os.Args[1:]))
}
